### ISSUE TITLE
Memory issues in 3.8

### ISSUE THREAD

**Issue by adamjmcgrath** [emberjs/ember.js#18140]
We're having memory retention issues upgrading from 3.4 to 3.8. When we use string based event listeners they get retained on the meta of the object that's being listened to. eg

```js
// "3.8.2"
var person = Ember.Object.extend(Ember.Evented, {
  greet() {
    this.trigger('greet');
  }
}).create();
var Foo = { hello() { console.log('hello') }  }
person.on('greet', Foo, 'hello')
person.greet()
// hello
Ember.meta(person)._listeners
 //  (4) ["greet", {…}, "hello", false]
person.off('greet', Foo, 'hello')
Ember.meta(person)._listeners
// [{event: "greet", target: {…}, method: "hello", kind: 2}] '_listeners' still has the event
```

After `person.off('greet', Foo, 'hello')` we would expect `Ember.meta(person)._listeners` to be an empty array - this is the behaviour we see in 3.4 and is also the behaviour we see when we use function listeners:

```js
person.on('greet', Foo, Foo.hello)
Ember.meta(person)._listeners
 //  (4) ["greet", {…}, "hello", false]
person.off('greet', Foo, Foo.hello)
Ember.meta(person)._listeners
// [] An empty array
```

It looks like the root cause is in a change that was added to retain these events, and instead give them a `kind` property of `ListenerFlags.REMOVE` (in https://github.com/emberjs/ember.js/pull/16874/files#diff-1c610ca34cd0e59afd9814e4e3b92806R471).

A later change was added to remove function listeners all together - presumably to fix issues with function listeners being retained (in https://github.com/emberjs/ember.js/pull/16923/commits/7164bd6c69f4c9b3ce33d129aac564dc1ec95920), but this was not done for string listeners.

This behaviour is causing large memory retention issues in our application and stopping us from upgrading. Should string listeners be cleaned up as well or should we change these all to function listeners to avoid the retention issues?

Have also tested this in 3.10 and see the same issue.

**Comment by adamjmcgrath** [comment]
I can also reproduce the same memory leak issue on https://try.discourse.org (they're on 3.8.0)

Detached HTML elements are being retained when toggling between the 'Top' and 'Categories' routes

![Screenshot 2019-06-25 at 16 26 02](https://user-images.githubusercontent.com/1299658/60111754-72cce600-9766-11e9-9bbf-33fee2f44338.png)

![Screenshot 2019-06-25 at 16 26 37](https://user-images.githubusercontent.com/1299658/60111778-7c564e00-9766-11e9-9026-86717e79c922.png)

**Comment by pzuraq** [comment]
Hey @adamjmcgrath, I worked on those changes so I think I can give a bit of context here. The refactors were done in order to make listeners _lazy_, there actually were some pretty serious bugs that could occur by using them in the wrong order, and these bugs were exacerbated by native classes.

Function listeners, in particular, were tricky because _inheriting_ them was hard. It didn't really make sense in most cases either. However, we did _not_ deprecate non-inherited function listeners (e.g. _instance_ listeners). Function listeners _will_ be fully removed too, as compared to string listeners, so you should be able to use switch to them to solve your problem in the short term.

In the longer term, I'm not sure if it's possible for us to remove string listeners on instances either. There are some funky edge cases where people may be using them for inheritance on POJOs, which was the reason we didn't before I believe. I'll talk with @krisselden about it to try to think through it again and figure out why we made that decision.

Also, I _do_ recommend in general moving away from listeners. They don't exist on Glimmer components and don't work particularly well with native classes in general (and Ember doesn't ship decorators for them) so I think minimizing their use would be a good idea.

**Comment by adamjmcgrath** [comment]
Hi @pzuraq, thanks for the explanation. It feels like bit of a gotcha that string listeners are removed lazily and function listeners aren't - albeit only from a performance standpoint. That said, it shouldn't be too hard for us to move away from listeners or at least change a few to function listeners in the interim.

FWIW, I noticed that LinkedIn also has the lazy removeListeners change in their version of Ember and I see detached HTML Elements being retained when I toggle to and from the Profile page.

<img width="1552" alt="Screenshot 2019-06-26 at 10 16 13" src="https://user-images.githubusercontent.com/1299658/60172870-7fece200-9805-11e9-804f-1803a33ec58f.png">
<img width="1438" alt="Screenshot 2019-06-26 at 10 07 10" src="https://user-images.githubusercontent.com/1299658/60172879-84b19600-9805-11e9-903a-3104c9020098.png">

**Comment by krisselden** [comment]
The intention was string listeners is what should be used for inheritance on prototypes and they should point to methods not functions with non static state and on instances function listeners added/removed should be removed. If the latter is not happening then it is a bug.

**Comment by krisselden** [comment]
Basically the only case that should leak is if you add a listener to a prototype that closes over instance state but there is no reasonable use case for that, inheriting listeners should only be for methods and use strings.  All add/remove of listeners to a non-prototype should not leak memory and it is a bug if they do.

**Comment by BillyRayPreachersSon** [comment]
Will this fix be backported to Ember 3.8, so that LTS users get the benefit? This is really hampering us upgrading from 3.4 LTS, and the memory usage in our app goes through the roof.

Thanks!
